### PR TITLE
Return an error instead of nil when parseline gets nil/empty input

### PR DESCRIPTION
**Pull request by ryancox**
This is a result of my first pass at go-fuzz'ing the lexer

### CODE REVIEW

File: statsd/receiver.go [hunk 1]
```diff
@@ -188,10 +189,15 @@ func (mr *metricReceiver) getAdditionalTags(addr string) types.Tags {
 
 // ParseLine with lexer impl.
 func (mr *metricReceiver) parseLine(line []byte) (*types.Metric, error) {
+	if line == nil {
+		return nil, errors.New("Error parsing line. Byte array is nil")
+	}
```
Comment (ash2k) [inline statsd/receiver.go]: I don't think this check is necessary, is it?

File: statsd/receiver.go [hunk 2]
```diff
@@ -188,10 +189,15 @@ func (mr *metricReceiver) getAdditionalTags(addr string) types.Tags {
...
+
 	llen := len(line)
 	if llen == 0 {
-		return nil, nil
+		return nil, errors.New("Error parsing line. Byte array is empty")
```
Comment (ash2k) [inline statsd/receiver.go]: I don't think it is possible to hit this branch in the current codebase - there is a check before this function is invoked:
```go
if len(line) > 1 {
    metric, err := mr.parseLine(line)
```
I agree the code is not ideal (a redundant check inside of the function) and I've cleanup it up already in another branch.

**Comment by ash2k** [comment]
Thanks for doing this! Can you share the input data corpus you used? Tooling/code you wrote? I'd like to include it in the build process.

**Comment by ryancox** [comment]
I have a WIP [fuzz branch](https://github.com/atlassian/gostatsd/compare/master...ryancox:fuzz) that includes a few items in the corpus as well as a new 'fuzz' target in the Makefile. Regarding the above questions about defensive checking, you will see if you look at that branch that there is code that does round-tripping using a call to parseFile. Without those checks, go-fuzz reports a crasher.

**Comment by ash2k** [comment]
I've merged my branch where this piece is refactored. Thanks for the effort! I'll take a look at your branch when I have time. Really keen to add fuzz testing to this project.

**Pull request closed without merging**